aarch64/kernel/nano_core/main.rs

```diff
@@ -30,16 +40,61 @@ fn main(
     uefi_services::init(&mut system_table)
         .map_err(|_| "nano_core::main - couldn't init uefi services")?;
 
-    let bootsvc = system_table.boot_services();
+    let boot_svc = system_table.boot_services();
 
     let safety = 16;
 
-    let mmap_size = bootsvc.memory_map_size();
+    let mmap_size = boot_svc.memory_map_size();
     let mut mmap = vec![0; mmap_size.map_size + safety * mmap_size.entry_size];
 
-    let _ = system_table.exit_boot_services(handle, &mut mmap)
+    let (_runtime_svc, mem_iter) = system_table.exit_boot_services(handle, &mut mmap)
         .map_err(|_| "nano_core::main - couldn't exit uefi boot services")?;
 
+    // Identifying free and reserved regions so we can initialize the frame allocator.
+    let mut free_regions: [Option<PhysicalMemoryRegion>; 32] = Default::default();
+    let mut free_index = 0;
+    let mut reserved_regions: [Option<PhysicalMemoryRegion>; 32] = Default::default();
+    let mut reserved_index = 0;
+
+    for descriptor in mem_iter {
+        let page_count = descriptor.page_count as usize;
+        let size = page_count * PAGE_SIZE;
+        if size > 0 {
+            let region_type = uefi_conv::convert_mem(descriptor.ty);
+            let (dst, index) = match region_type {
+                MemoryRegionType::Free => (&mut free_regions, &mut free_index),
+                MemoryRegionType::Reserved => (&mut reserved_regions, &mut reserved_index),
+                MemoryRegionType::Unknown => continue,
+            };
+
+            let start_addr = descriptor.phys_start as usize;
+            let start_addr = PhysicalAddress::new_canonical(start_addr);
+            let end_addr = start_addr + size;
+
+            let first_frame = Frame::containing_address(start_addr);
+            let last_frame = Frame::containing_address(end_addr - 1);
+
+            let range = FrameRange::new(first_frame, last_frame);
```

Comment:
there's a convenience function for this that does what you want: https://www.theseus-os.com/Theseus/doc/memory_structs/struct.FrameRange.html#method.from_phys_addr

```suggestion
            let range = FrameRange::from_phys_addr(start_addr, size);
```